lib/SIL/Verifier/LinearLifetimeChecker.cpp

```diff
@@ -307,7 +307,13 @@ void State::checkForSameBlockUseAfterFree(Operand *consumingUse,
   for (auto *nonConsumingUse : nonConsumingUsesInBlock) {
     if (std::find_if(consumingUse->getUser()->getIterator(), userBlock->end(),
                      [&nonConsumingUse](const SILInstruction &i) -> bool {
-                       return nonConsumingUse->getUser() == &i;
+                       if (nonConsumingUse->getUser() != &i)
+                         return false;
+                       // Reborrows are currently represented as a consuming
+                       // use. Do not raise error for reborrows.
+                       if (isReborrowInstruction(&i))
```

Comment:
I thought about this in a deeper manner. The condition we are really looking for here is if the consuming use and the non consuming use are the same instruction. We can just hoist that out of the if into an earlier if statement and continue. I.e.:

```
for (auto *nonConsumingUse : nonConsumingUsesInBlock) {
  auto *consumingUser = consumingUse->getUser();
  if (nonConsumingUse->getUser() != consumingUser)) {
     if (std::find_if(... normal code ...)) {
        continue;
     }
  } else {
    if (isReborrowInstruction(consumingUser))
       continue;
  }

  ... throw error ...
}
```

Reply:
I am also worried about baking in isReborrowInstruction into the verifier itself. It is meant to be a more general utility than this. Maybe add a flag that controls whether or not we allow for reborrows and have it disabled by default? Then at least we are only making this change where it is needed.

Another thing, on the reborrow failure path we could emit a better error as well (since it is not exactly an outside of lifetime use).

Reply:
I refactored. But am not sure why would we need a flag ? Reborrows are never consuming